core/server/master/src/main/java/alluxio/master/meta/PathProperties.java

```diff
@@ -12,41 +12,235 @@
+      mState.applyAndJournal(ctx, PathPropertiesEntry.newBuilder()
+          .setPath(path).putAllProperties(newProperties).build());
+    }
   }
 
   /**
-   * Sets a property for a path pattern.
+   * Removes the specified set of keys from the properties for path.
    *
-   * @param path the path pattern
-   * @param key the property key
-   * @param value the property value
+   * @param ctx the journal context
+   * @param path the path
+   * @param keys the keys to remove
+   */
+  public void remove(Supplier<JournalContext> ctx, String path, Set<PropertyKey> keys) {
+    try (LockResource r = new LockResource(mLock.writeLock())) {
+      if (mState.getProperties().containsKey(path)) {
+        Map<String, String> newProperties = mState.getProperties().get(path);
+        keys.forEach(key -> newProperties.remove(key.getName()));
+        mState.applyAndJournal(ctx, PathPropertiesEntry.newBuilder()
+            .setPath(path).putAllProperties(newProperties).build());
+      }
+    }
+  }
+
+  /**
+   * Removes all properties for path.
+   *
+   * @param ctx the journal context
+   * @param path the path
+   */
+  public void removeAll(Supplier<JournalContext> ctx, String path) {
+    try (LockResource r = new LockResource(mLock.writeLock())) {
+      if (mState.getProperties().containsKey(path)) {
+        mState.applyAndJournal(ctx, RemovePathPropertiesEntry.newBuilder()
+            .setPath(path).build());
+      }
+    }
+  }
+
+  @Override
+  public Journaled getDelegate() {
+    return mState;
+  }
+
+  /**
+   * Journaled state of path level properties.
    */
-  public void setProperty(String path, PropertyKey key, String value) {
-    mProperties.computeIfAbsent(path, k -> new ConcurrentHashMap<>()).put(key, value);
+  @NotThreadSafe
+  public static final class State implements Journaled {
+    /**
+     * Map from path to key value properties.
+     */
+    private final Map<String, Map<String, String>> mProperties = new HashMap<>();
+
+    /**
+     * The map from paths to properties is unmodifiable, but for each path, the property map is
+     * modifiable.
+     *
+     * @return the internal properties
+     */
+    public Map<String, Map<String, String>> getProperties() {
+      return Collections.unmodifiableMap(mProperties);
+    }
+
+    @Override
+    public boolean processJournalEntry(Journal.JournalEntry entry) {
+      if (entry.hasPathProperties()) {
+        applyPathProperties(entry.getPathProperties());
+      } else if (entry.hasRemovePathProperties()) {
+        applyRemovePathProperties(entry.getRemovePathProperties());
+      } else {
+        return false;
+      }
+      return true;
+    }
+
+    private void applyPathProperties(PathPropertiesEntry entry) {
+      final String path = entry.getPath();
+      final Map<String, String> properties = entry.getPropertiesMap();
+      if (mProperties.containsKey(path)) {
+        mProperties.get(path).clear();
+        mProperties.get(path).putAll(properties);
```

Comment:
Is it necessary to keep the previous hashmap, could we just use put which will override?

Reply:
`properties` is an unmodifiable map

Reply:
I meant - instead of the `if-else`, could we just do `mProperties.put(path, new HashMap<>(properties))`?

Reply:
the internal implementation of `new HashMap(Map)` is actually the same as `putAll`, but I don't want to create unnecessary new object instances.

Reply:
Got it, thanks. Just curious, what are the respective overheads (clear + putAll) vs make new object?